// codeNamesOnline/database/databaseManager.js
'use strict';
//==============================================================================
//  Modules
//==============================================================================

var databaseConfig = require('codeNamesOnline/configManager').LoadConfig('database');
var knex           = require('knex')({
    client      : databaseConfig.client,
    connection  : {
        socketPath  : databaseConfig.socketPath,
        host        : databaseConfig.host,
        user        : databaseConfig.user,
        password    : databaseConfig.password,
        database    : databaseConfig.database,
    },
});

//==============================================================================
//  Public functions
//==============================================================================

/*
 *  public object GetQueryBuilder()
 *
 *  Returns the query builder used to execute SQL queries
 *
 *  @return {object} query builder
 */
module.exports.GetQueryBuilder = function () {
	return knex;
};
