// codeNamesOnline/game.js
'use strict';

//==============================================================================
//  Modules
//==============================================================================

var socketManager = require('codeNamesOnline/socketManager');
var wordType      = require('codeNamesOnline/constants/wordType');

//==============================================================================
//  Exports
//==============================================================================

module.exports = function (gameCode, wordGrid, startTeam, socketId) {
	var self               = this;
	var gameCode           = gameCode;
	var wordGrid           = wordGrid;
	var socketId           = socketId;
	var turn               = startTeam;
	var blueTeamWordsLeft  = null;
	var redTeamWordsLeft   = null;
	var blueTeamCodeMaster = null;
	var redTeamCodeMaster  = null;

	if (startTeam === 'blue') {
		blueTeamWordsLeft = 9;
		redTeamWordsLeft = 8;
	}
	else {
		blueTeamWordsLeft = 8;
		redTeamWordsLeft = 9;
	}

	// Getters

	this.getGameCode = function () {
		return gameCode;
	};

	this.getBlueTeamCodeMaster = function () {
		return blueTeamCodeMaster;
	};

	this.getRedTeamCodeMaster = function () {
		return redTeamCodeMaster;
	};

	this.getTurn = function () {
		return turn;
	};

	// Setters

	this.setBlueTeamCodeMaster = function (codeMaster) {
		if (blueTeamCodeMaster !== null) {
			throw new Error('This game already has a Code Master for Blue Team');
		}
		blueTeamCodeMaster = codeMaster;
		AddPlayer(blueTeamCodeMaster);
	};

	this.setRedTeamCodeMaster = function (codeMaster) {
		if (redTeamCodeMaster !== null) {
			throw new Error('This game already has a Code Master for Red Team');
		}
		redTeamCodeMaster = codeMaster;
		AddPlayer(redTeamCodeMaster);
	};

	this.removeBlueTeamCodeMaster = function () {
		blueTeamCodeMaster = null;
	};

	this.removeRedTeamCodeMaster = function () {
		redTeamCodeMaster = null;
	};

	// Public functions

	this.getGameData = function () {
		return {
			wordGrid          : wordGrid,
			blueTeamWordsLeft : blueTeamWordsLeft,
			redTeamWordsLeft  : redTeamWordsLeft,
			turn              : turn
		};
	};

	this.selectWord = function (word) {
		// TODO Search grid for work and set chosen to true
		// Score game here too, no need for the setters above
		UpdateGame();
	};

	this.nextTurn = function () {
		if (turn === 'blue') {
			turn = 'red';
		}
		else {
			turn = 'blue';
		}
		UpdateGame();
	};

	this.EndGame = function () {
		// TODO send message to game and players ending the game.
	};

	// Private functions

	function UpdateGame () {
		var gameSocket = socketManager.GetSocket(socketId);
		if (gameSocket) {
			gameSocket.emit('updateGameData', self.getGameData);
		}
		else {
			console.log('Unable to update game! No socket');
		}
	}

	// Tell the frontend that a player has been added
	function AddPlayer (player) {
		var gameSocket = socketManager.GetSocket(socketId);
		gameSocket.emit('playerConnect', {
			name : player.getName(),
			team : player.getTeam()
		});
	}
};
