// codeNamesOnline/gameManager.js
'use strict';

//==============================================================================
//  Modules
//==============================================================================

var Game              = require('codeNamesOnline/game');
var Player            = require('codeNamesOnline/player');
var wordGridGenerator = require('codeNamesOnline/wordGridGenerator');
var wordType          = require('codeNamesOnline/constants/wordType');

//==============================================================================
//  Games
//==============================================================================

var currentGames = [];

//==============================================================================
//  Exports
//==============================================================================

module.exports.CreateNewGame = function (callback) {
	var gameCode = GenerateGameCode();
	wordGridGenerator.generateWordGrid(function (error, wordGrid) {
		if (error) {
			return callback(error);
		}
		else {
			currentGames.push(new Game(gameCode, wordGrid));
			return callback(null, gameCode);
		}
	});
};

module.exports.AddSocket = function (gameCode, socket) {
	var game = GetGame(gameCode);
	if (game) {
		game.setSocket(socket);
	}
	else {
		throw new Error('Cannot add socket. Game \'' + gameCode + '\' does not exist');
	}
};

module.exports.RemoveGame = function (gameCode) {
	for (var i = 0; i < currentGames.length; i++) {
		if (currentGames[i].getGameCode() === gameCode) {
			currentGames[i].EndGame();
			currentGames.splice(i, 1);
		}
	}
	throw new Error('Cannot remove game. Game \'' + gameCode + '\' does not exist');
};

module.exports.AddPlayer = function (gameCode, player) {
	var game = GetGame(gameCode);
	if (game) {
		if (player.getTeam() === wordType.BLUE_TEAM) {
			game.setBlueTeamCodeMaster(player);
		}
		else if (player.getTeam() === wordType.RED_TEAM) {
			game.setRedTeamCodeMaster(player);
		}
		else {
			throw new Error('Cannot add player. Team \''
			+ player.getTeam() + '\' does not exist');
		}
	}
	else {
		throw new Error('Cannot add player. Game \'' + gameCode + '\' does not exist');
	}
};

module.exports.RemovePlayer = function (gameCode, team) {
	var game = GetGame(gameCode);
	if (game) {
		if (team === wordType.BLUE_TEAM) {
			game.removeBlueTeamCodeMaster();
		}
		else if (team === wordType.RED_TEAM) {
			game.removeRedTeamCodeMaster();
		}
		else {
			throw new Error('Cannot remove player. Team \''
			+ team + '\' does not exist');
		}
	}
	else {
		throw new Error('Cannot remove player. Game \'' + gameCode + '\' does not exist');
	}
};

//==============================================================================
//  Private functions
//==============================================================================

function GetGame (gameCode) {
	for (var i = 0; i < currentGames.length; i++) {
		if (currentGames[i].getGameCode() === gameCode) {
			return currentGames[i];
		}
	}
	return null;
}

function GenerateGameCode () {

	var gameCode;
	var possible = 'abcdefghijklmnopqrstuvwxyz';

	do {
		gameCode = '';
		for (var i = 0; i < 4; i++) {
			gameCode += possible.charAt(Math.floor(Math.random() * possible.length));
		}
	} while(!isUniqueGameCode(gameCode));

	return gameCode;
}

function isUniqueGameCode (gameCode) {
	for (var i = 0; i < CurrentGames.length; i++) {
		if (gameCode === CurrentGames[i].GetGameCode()) {
			return false;
		}
	}
	return false;
}
