// codeNamesOnline/game.js
'use strict';

module.exports = function (gameCode, wordGrid) {
	var self               = this;
	var gameCode           = gameCode;
	var wordGrid           = wordGrid;
	var socket             = null;
	var blueTeamScore      = 0;
	var redTeamScore       = 0;
	var blueTeamCodeMaster = null;
	var redTeamCodeMaster  = null;
	var turn               = 'blue';

	// Getters

	this.getGameCode = function () {
		return gameCode;
	};

	this.getWordGrid = function () {
		return wordGrid;
	};

	this.getSocket = function () {
		return socket;
	};

	this.getBlueTeamScore = function () {
		return blueTeamScore;
	};

	this.getRedTeamScore = function () {
		return redTeamScore;
	};

	this.getBlueTeamCodeMaster = function () {
		return blueTeamCodeMaster;
	};

	this.getRedTeamCodeMaster = function () {
		return redTeamCodeMaster;
	};

	this.getTurn = function () {
		return turn;
	};

	// Setters

	this.setSocket = function (newSocket) {
		socket = newSocket;
	};

	this.setBlueTeamCodeMaster = function (codeMaster) {
		if (blueTeamCodeMaster !== null) {
			throw new Error('This game already has a Code Master for Blue Team');
		}
		blueTeamCodeMaster = codeMaster;
		UpdateGame();
	};

	this.setRedTeamCodeMaster = function (codeMaster) {
		if (redTeamCodeMaster !== null) {
			throw new Error('This game already has a Code Master for Red Team');
		}
		redTeamCodeMaster = codeMaster;
		UpdateGame();
	};

	this.removeBlueTeamCodeMaster = function () {
		blueTeamCodeMaster = null;
		UpdateGame();
	};

	this.removeRedTeamCodeMaster = function () {
		redTeamCodeMaster = null;
		UpdateGame();
	};

	// Public functions

	this.getGameData = function () {
		return {
			wordGrid      : wordGrid,
			blueTeamScore : blueTeamScore,
			redTeamScore  : redTeamScore,
			turn          : turn
		};
	};

	this.selectWord = function (word) {
		// TODO Search grid for work and set chosen to true
		// Score game here too, no need for the setters above
		UpdateGame();
	};

	this.nextTurn = function () {
		if (turn === 'blue') {
			turn = 'red';
		}
		else {
			turn = 'blue';
		}
		UpdateGame();
	};

	this.EndGame = function () {
		// TODO send message to game and players ending the game.
	};

	// Private functions

	function UpdateGame () {
		// TODO send message to game and players updating the game state.
		// self.getGameData()
	}
};
