// codeNamesOnline/api/gameAPI.js
'use strict';

//==============================================================================
//  Modules
//==============================================================================

var gameManager = require('codeNamesOnline/gameManager');
var Player      = require('codeNamesOnline/player');

//==============================================================================
//  Exports
//==============================================================================


module.exports.CreateNewGame = function (socketId, callback) {
	gameManager.CreateNewGame(socketId, function (error, gameCode) {
		if (error) {
			return callback(error);
		}
		else {
			return callback(null, gameCode);
		}
	});
};

module.exports.GetGameData = function (gameCode, callback) {
	try {
		var gameData = gameManager.GetGameData(gameCode);
	}
	catch (ex) {
		return callback(ex);
	}
	return callback(null, gameData);
};

module.exports.AddPlayer = function (gameCode, playerDataJSON, socketId, callback) {
	try {
		var playerData = JSON.parse(playerDataJSON);
		var name = playerData.name;
		var team = playerData.team;
		var player = new Player(name, team, socketId);
		gameManager.AddPlayer(gameCode, player);
	}
	catch (ex) {
		return callback(ex);
	}
	return callback();
};

module.exports.SelectWord = function (gameCode, word, callback) {
	try {
		gameManager.SelectWord(gameCode, word);
	}
	catch (ex) {
		return callback(ex);
	}
	return callback();
};

module.exports.NextTurn = function (gameCode, callback) {
	try {
		gameManager.NextTurn(gameCode);
	}
	catch (ex) {
		return callback(ex);
	}
	return callback();
};

module.exports.ResetGame = function (gameCode, callback) {
	gameManager.ResetGame(gameCode, function (error) {
		if (error) {
			return callback(error);
		}
		else {
			return callback();
		}
	});
};
